src/vivarium_census_prl_synth_pop/tools/make_results.py

```diff
@@ -548,3 +552,67 @@ def subset_results_by_state(processed_results_dir: str, state: str) -> None:
             )
             write_to_disk(state_data, output_file_path)
         logger.info(f"Finished writing {observer} files for {state_name}.")
+
+
+def write_shard_metadata(
+    observer: str, obs_data: pd.DataFrame, obs_dir: Path, seed: str
+) -> None:
+    # Writes metadata for each shard of data. This will be proportions available to noise
+    # for specific columns.
+
+    def _get_metadata_values(df: pd.DataFrame, location: str, year: int) -> pd.DataFrame:
+        metadata_df = pd.DataFrame(index=[location])
+        metadata_df["number_of_rows"] = len(df)
+        metadata_df["year"] = year
+        copy_columns = [
+            col.name
+            for col in COLUMNS
+            for noise_type in col.noise_types
+            if "copy_from_household_member" == noise_type.name
+        ]
+        nicknames_columns = [
+            col.name
+            for col in COLUMNS
+            for noise_type in col.noise_types
+            if "use_nickname" == noise_type.name
+        ]
+        for column in df.columns:
+            if column in copy_columns:
+                # Get number of rows that could potentially copy a household member
+                metadata_df[f"{column}.copy_from_household_member"] = (
+                    df[COPY_HOUSEHOLD_MEMBER_COLS[column]].notna().sum()
+                )
+            elif column in nicknames_columns:
+                # Get number of rows eligible to be noised to a nickname
+                nicknames = load_nicknames_data()
+                metadata_df[f"{column}.use_nickname"] = df[column].isin(nicknames.index).sum()
+            else:
+                # TODO: Add guardian based duplication
+                continue
+        return metadata_df
+
+    # Get year column to group by
+    date_columns = ["year", "tax_year", "event_date", "survey_date"]
+    year_col = [col for col in obs_data.columns if col in date_columns]
+    state_col = "state" if "state" in obs_data.columns else "mailing_address_state"
+    # Special case SSA dataset since that does not have a state column
+    if observer == metadata.DatasetNames.SSA:
+        metadata_dfs = []
+        for year, year_data in obs_data.groupby(year_col):
+            year_df = obs_data.loc[year_data.index]
+            year_metadata = _get_metadata_values(year_df, "USA", year)
+            metadata_dfs.append(year_metadata)
+    else:
+        groupby_cols = year_col + [state_col]
+        metadata_dfs = []
+        for (year, location), location_data in obs_data.groupby(groupby_cols):
+            state_df = obs_data.loc[location_data.index]
+            state_metadata = _get_metadata_values(state_df, location, year)
+            metadata_dfs.append(state_metadata)
```

Comment:
This could maybe be simplified:

```python
if observer == metadata.DatasetNames.SSA:
    obs_data[state_col] = "USA"

metadata_dfs = []
for (year, location), group_data in obs_data.groupby(groupby_cols):
    group_data = obs_data.loc[group_data.index]    # I think this is not actually doing anything? See question above.
    group_metadata = _get_metadata_values(group_data , location, year)
    metadata_dfs.append(group_metadata)
```